tests/native_test_native.h

```diff
@@ -0,0 +1,276 @@
+// automatically generated by the FlatBuffers compiler, do not modify
+
+
+#ifndef FLATBUFFERS_NATIVE_NATIVETEST_TESTN_H_
+#define FLATBUFFERS_NATIVE_NATIVETEST_TESTN_H_
+
+#include "native_test_generated.h"
+#include "custom_native.h"
+
+#include "flatbuffers/flatbuffers.h"
+#include <variant>
+
+namespace TestN {
+namespace Native {
+
+struct DefindeBeforeUnion;
+struct Foo;
+}  // namespace Native
+
+namespace NamespaceFoo {
+namespace Native {
+
+struct Foo;
+}  // namespace Native
+}  // namespace NamespaceFoo
+
+namespace NamespaceBar {
+namespace Native {
+
+struct Foo;
+}  // namespace Native
+}  // namespace NamespaceBar
+
+namespace Native {
+
+struct MyUnionUnion;
+
+bool operator==(const DefindeBeforeUnion &lhs, const DefindeBeforeUnion &rhs);
+
+bool operator==(const Foo &lhs, const Foo &rhs);
+
+}  // namespace Native
+
+namespace NamespaceFoo {
+namespace Native {
+
+bool operator==(const Foo &lhs, const Foo &rhs);
+
+}  // namespace Native
+}  // namespace NamespaceFoo
+
+namespace NamespaceBar {
+namespace Native {
+
+bool operator==(const Foo &lhs, const Foo &rhs);
+
+}  // namespace Native
+}  // namespace NamespaceBar
+
+namespace Native {
+
+struct DefindeBeforeUnion : public flatbuffers::NativeTable {
+  typedef ::TestN::DefindeBeforeUnion TableType;
+  int32_t dummy;
+  DefindeBeforeUnion()
+      : dummy(0) {
+  }
+};
+
+inline bool operator==(const DefindeBeforeUnion &lhs, const DefindeBeforeUnion &rhs) {
+  return
+      (lhs.dummy == rhs.dummy);
+}
+
+struct MyUnionUnion : public std::variant<flatbuffers::NoneType, MyMat, DefindeBeforeUnion> {
+    using std::variant<flatbuffers::NoneType, MyMat, DefindeBeforeUnion>::variant;
+};
+
+struct Foo : public flatbuffers::NativeTable {
+  typedef ::TestN::Foo TableType;
+  ::TestN::BundleSize enumData;
+  MyMat bitData;
+  std::vector<std::complex<double>> iqData;
```

Comment:
@sketch34 related to https://github.com/google/flatbuffers/issues/4969#issuecomment-1027646847. Yes, using this native type extension, you can do what you asked for. Simply specify "--cpp-ptr-type inline". Unfortunately, adapting and improving the current upstream code is challenging for me. I instead simply use two code generators. First the vanilla one, second a "native_type" generator with the code presented in this PR